core/src/main/scala/filodb.core/memstore/TimeSeriesPartition.scala

```diff
@@ -392,6 +392,34 @@ extends ChunkMap(memFactory, initMapSize) with ReadablePartition {
   }
 }
 
+/**
+ * A variant of the above which logs every sample ingested and buffer switching/encoding event,
+ * for debugging purposes.
+ */
+class TracingTimeSeriesPartition(partID: Int,
+                                 dataset: Dataset,
+                                 partitionKey: BinaryRegion.NativePointer,
+                                 shard: Int,
+                                 bufferPool: WriteBufferPool,
+                                 shardStats: TimeSeriesShardStats,
+                                 memFactory: MemFactory,
+                                 initMapSize: Int) extends
+TimeSeriesPartition(partID, dataset, partitionKey, shard, bufferPool, shardStats, memFactory, initMapSize) {
```

Comment:
Prefer composition /  decorator pattern over inheritance.  
Inheritance leads to bad problems later on (remember Stack extends from Vector in java) ? One such place we already have  in code is OnDemandPagingShard which should have not extended TimeSeriesShard.

However, I dont have an alternate solution with less code changes without adding more memory usage for TSP. :(

Please add a comments so readers understand.

Reply:
The reason why I used inheritance was not so much memory but just ease of implementation.  With composition we'd need to add in tons of methods and clutter things up quite a bit.  If it simply implemented `ReadablePartition` that might break things in a bunch of places.  So best way to keep changes small and balance out different needs.